core/che-core-api-dto/pom.xml

```diff
@@ -46,6 +46,10 @@
             <groupId>org.eclipse.che.core</groupId>
             <artifactId>che-core-commons-lang</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.lsp4j</groupId>
+            <artifactId>org.eclipse.lsp4j.jsonrpc</artifactId>
```

Comment:
I'm kind of -1 for this dependency. DTO is supposed to be feature agnostic.
Can we avoid to use it here? Maybe we should make dto mechanism extensible so you can override it?

Reply:
(Answering this and next comment together.)

We need this because at the moment Dto generation for LSP4J handles the Either class with special handling. It "looks inside" Eithers and stores the active one. This is compatible with how LSP4J serializes and deserializes, but only if EitherTypeAdapterFactory is used. 

AFAICT, based on the contents of #3103, originally Dto handled the serialize and deserialize all the time, i.e. whatever Dto wrote to the stream, Dto would read from the stream so Dto only had to be consistent with itself. But #3103 added using Gson as an option to doing the deserialize (depending on which variant of DtoFactory.createDtoFromJson is called). 

The alternative solutions to this problem:

- Don't use Gson for deserialize when doing LSP4J (is there the ability to provide this fine grained control of which createDtoFromJson is called?)
- Change LSP4J Dto creation to stop handling LSP4J's Either in a special way (I don't know if this is possible, can Gson handle the semantics of Either without an adapter?)
- Make it extensible (E.g. a method to register additional adapter factories). for example register adapter factory at the same time as registering providers. If new factories are added the gson instance has to be replaced, or the gson instance creation has to be delayed until all factories are registered

Reply:
>Make it extensible

See my comment below.